### ISSUE TITLE
Image only part of the data

### ISSUE THREAD

**Issue by AHorneffer** [lofar-astron/prefactor#97]
To subtract a facet Factor does not image the full bandwidth anymore, but only images some bands spread out over the full BW and then fits a spectral index.
This could also be used with the `Initial-Subtract` pipeline.

It is probably good to implement that together with:  #56

**Comment by darafferty** [comment]
I added a version of the `Initial-Subtract` pipeline (called `Initial-Subtract-Fast`) to the experimental branch that implements this. By default, it images 6 bands spread across the full bandwidth and interpolates the sky models between them. Additionally, it images just once per band and high and low resolution and uses the mask to filter the clean components (instead of imaging twice and using the mask as a clean mask). Lastly, it supports PBS schedulers, so can distribute the work across multiple nodes of a cluster.

Things to watch out for are: whether the interpolation of the sky models works OK, whether the single imaging run is good enough, and whether the script that makes sky models needs further tuning (e.g., it neglects CCs below 5 mJy to keep the sky models from growing too large, and this may be too aggressive).

**Comment by Joshuaalbert** [comment]
What type of interpolation do you use? Tri cubic would maintain C1 properties.On 13 Jan 2017 21:57, David Rafferty <notifications@github.com> wrote:I added a version of the Initial-Subtract pipeline (called Initial-Subtract-Fast) to the experimental branch that implements this. By default, it images 6 bands spread across the full bandwidth and interpolates the sky models between them. Additionally, it images just once per band and high and low resolution and uses the mask to filter the clean components (instead of imaging twice and using the mask as a clean mask). Lastly, it supports PBS schedulers, so can distribute the work across multiple nodes of a cluster.
Things to watch out for are: whether the interpolation of the sky models works OK, whether the single imaging run is good enough, and whether the script that makes sky models needs further tuning (e.g., it neglects CCs below 5 mJy to keep the sky models from growing too large, and this may be too aggressive).

**Comment by darafferty** [comment]
It uses simple linear interpolation. If you want to experiment with other ones (any supported by `scipy.interpolate.interp1d` should work), you can set the method in the parset by adding it to the flags in the `fits_to_bbs` steps. E.g.:

`fits_to_bbs_high.argument.flags           =   [rootname,msfile,skymodel,mask,cubic]`

**Comment by Joshuaalbert** [comment]
@darafferty See my edited previous comment

**Comment by darafferty** [comment]
Maybe you could just replace the call to `scipy.interpolate.interp1d()` in the script (line 159 of https://github.com/lofar-astron/prefactor/blob/experimental/scripts/fits2sky.py) with your method. It doesn't need to support vectorized input (the interpolation is done separately for each source).

**Comment by Joshuaalbert** [comment]
I see that above there you first look for non-zero stackings of models and masks and then perform interpolation at those points along freq axis. In this way each pixel is independent of its neighbours pixels. This is probably fine if clean component astrometry is sub-pixel, then scipy's cubic interpolator is fine which is the choice you make already. Now tricubic would only be useful if you wanted to use interpolation in 3d so including the neighbouring in pixels. From experience looking at clean components often a source is a splotch of model points sometimes with holes. There is clearly spatial coherence over the sources to some extent. Basically if we want to add tricubic interpolation here then I need three vectors and a 3-array. Example:
```
import TricubicInterpolation as TCI

#get freq axis, get pixel axes and the scalar points in a 3-tensor. 
#They must be regular be not necessarily uniform.
#so the 3-tensor is a defined over the grid
#e.g.
freqVec = [0,1,3,4,5]
raVec = [0,1,2,4,5,6,7]
decVec = [4,6,7]

fluxArray = np.ones([len(freqVec), len(raVec), len(decVec)])
tci = TCI.TriCubic(freqVec, raVec, decVec, fluxArray)

#then loop over all points where you want interpolated fluxes and do:
f = tci.interp1(freq,ra,dec)
#Never interpolate near edges of freqvec,ravec, or decvec.
#Currently, you only interpolate along freq where models have at least one non-zero freq, AND masks have at least one True.
```

**Comment by AHorneffer** [comment]
I guess one significant difference between the interpolation here and in Factor is that in Factor there is one call to WSClean, using the `channelsout` feature/parameter of WSClean. This way WSClean makes sure that there are the "same" clean-components at the same pixel positions in all channel-images.
But if I see it correctly, then in prefactor the imaging for the different channels is done in separate WSClean calls. Thus it is indeed possible that a source is distributed differently over the pixels in the different channel-images.
@darafferty Is that correct?

**Comment by darafferty** [comment]
Yes, that's right -- so a pixel could have fluxes/gaps at different frequencies than its neighbors do (not sure if this matters for the 3D interpolation or not, though)

**Comment by Joshuaalbert** [comment]
Holes (zeros) do not matter to interpolation. As long as the point is within the bounds of the specified domain, then the interpolator will guaranty that the function matches at the given points (so it will reproduce  zeros where the models have them, but it will also give continuous derivatives. So the function with be smooth. The only thing it requires are the definitions of the three axes and the known values at the grid points. For this I would question first, does the wcs give recitlinear coordinates in (ra,dec). I believe this is a no. Thus the coordinates I would use are pixel indices. These are guarantied to be rectilinear. I would consider which pixels to interpolate. It seems logical that one could use the current scheme of choosing pixels that have at least one positive model component with a mask.

**Comment by Joshuaalbert** [comment]
Consider the 2d picture below. Linear interpolation will make the channel with b and g almost all zero ignoring the values of neighbouring cells. Now if d and e are substantial values and the reason there is a zero between d and e is because of discretization then we can use the information from neighbouring cells to help fill in the blanks. The information used is basically that such that derivatives are smooth.

```
|a|b|c|
| | | |
|d|0|e|
| | | |
|f|g|h|
```

**Comment by darafferty** [comment]
I see. It might indeed be useful to limit the effects of discretization, as I would think they result in a worse subtraction (and higher residuals).
